### DISCUSSION TITLE
schedule function syntax

### DISCUSSION THREAD

**Discussion by krishnak** [micropython/micropython#9396] [Using MicroPython] [answered]
`            schedule(value.performAction(),None)
`

I am trying to schedule a call to a interrupt handler like this.
The method performAction() doesn't take any arguments  it has a method signature like def performAction(self):

The interrupt occurs, the method is executed - all is well but on the console, I get a 
`TypeError: 'NoneType' object isn't callable
`
This error is not from the performAction method but from the call schedule - can you let me know what is the corrected syntax to avoid this error.

**Comment by jimmo** [comment, marked as answer]
Sounds like you want

`schedule(value.performAction,None)`

**Reply by jimmo** [reply]
What's happening in your case is that you're executing the function and passing it's return value (None) to schedule.

Instead you want to pass the actual function and schedule will call it for you.

**Reply by krishnak** [reply]
Got it! thanks

**Comment by krishnak** [comment]
I need to call this function from an ISR as a schedule
    async def delayedEvent(self,eventorder,isremoteEvent):

I have tried 
schedule(self.delayedEvent,(eventNumber,True))  Complains method takes 3 arguments but 2 were given
schedule(self.delayedEvent,(self,eventNumber,True))  Complains method takes 3 arguments but 2 were given

Can you please suggest the correct syntax

**Reply by peterhinch** [reply]
Please see my response to your other query. There is really only one way to interface `uasyncio` to interrupts, and that is to use `ThreadSafeFlag`.

**Reply by jimmo** [reply]
Yep, exactly what @peterhinch said.

`schedule` does not know how to run a coroutine (i.e. async function), it can only call regular functions.

And just to clarify, the reason for the error about the number of arguments is that the second argument to schedule is not a tuple of arguments. Rather it's just the single argument to the function. A function that is passed to schedule must take a single argument.

```
def foo(a):
  print(a)

schedule(foo, (1,2,3))
```

In this case, foo is passed a single argument a, which is a tuple containing three values.
